### PR TITLE
DPE-65 basic relation departed

### PR DESCRIPTION
**Pull request by MiaAltieri**
This PR addresses a basic relation departed. In which the administrator plans for a unit to leave with `juju remove-unit <unit name>`.

_Note this PR does not address the following_:
1. The departing unit is the leader
2. The departing unit is the primary
3. The departing unit departs unexpectedly with the possibility of becoming live again

The aforementioned list will be addressed in subsequent PRs.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -47,9 +47,27 @@ def __init__(self, *args):
...
         self.framework.observe(self.on.update_status, self._on_update_status)
         self.framework.observe(self.on.get_primary_action, self._on_get_primary_action)
 
+    def _on_mongodb_relation_departed(self, event: ops.charm.RelationDepartedEvent) -> None:
+        """Removes the unit from the MongoDB replica set.
+
+        Args:
+            event: The triggering relation departed event.
+        """
+        # only leader should configure replica set and app-changed-events can trigger the relation
```
Comment (MiaAltieri) [inline src/charm.py:62]: app-changed? Remove this comment since this is relation departed . Also no need to check even.unit

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -144,25 +144,25 @@ async def test_get_primary_action(ops_test: OpsTest) -> None:
 
 async def test_cluster_is_stable_after_deletion(ops_test: OpsTest) -> None:
     """Tests that the cluster maintains a primary after the primary is delelted."""
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:146]: after non-primary, non-leader

**Review by MiaAltieri — COMMENTED** [review]
minor comments

File: src/charm.py [hunk 3]
```diff
@@ -47,9 +47,27 @@ def __init__(self, *args):
...
+        # changed hook resulting in no JUJU_REMOTE_UNIT if this is the case we should return
+        print("Before reconfig", self._replica_set_hosts)
```
Comment (MiaAltieri) [inline src/charm.py:64]: remove print

File: src/charm.py [hunk 4]
```diff
@@ -47,9 +47,27 @@ def __init__(self, *args):
...
+        if not (self.unit.is_leader() and event.unit):
+            return
+
+        self._reconfigure(event)
+        print("After reconfig", self._replica_set_hosts)
```
Comment (MiaAltieri) [inline src/charm.py:69]: remove print

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -143,26 +143,26 @@ async def test_get_primary_action(ops_test: OpsTest) -> None:
 
 
 async def test_cluster_is_stable_after_deletion(ops_test: OpsTest) -> None:
-    """Tests that the cluster maintains a primary after the primary is delelted."""
-    # find & destroy leader unit
-    leader_ip = None
+    """Tests that the cluster cluster behavior after planned non-leader unit removal."""
```
Comment (marceloneppel) [inline tests/integration/test_charm.py:146]: Word `cluster` looks duplicated here.

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -175,14 +175,24 @@ async def test_cluster_is_stable_after_deletion(ops_test: OpsTest) -> None:
     except RetryError:
         primary = None
 
-    client.close()
-
     # verify that the primary is not None
     assert primary is not None, "replica set with uri {} has no primary".format(replica_set_uri)
 
     # check that the primary is not one of the deleted units
     assert primary[0] in ip_addresses, "replica set primary is not one of the available units"
```
Comment (marceloneppel) [inline tests/integration/test_charm.py:182]: This is not part of the change list, but I saw that the error message is not consistent with the check and the above comment.

**Review by marceloneppel — APPROVED** [review]
LGTM, Mia. Nice work. I left only two notes about comments.

File: src/charm.py [hunk 7]
```diff
@@ -47,9 +47,24 @@ def __init__(self, *args):
...
+        """Removes the unit from the MongoDB replica set.
+
+        Args:
+            event: The triggering relation departed event.
+        """
+        # only leader should configure replica set
+        if not self.unit.is_leader():
+            return
+
+        self._reconfigure(event)
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:66]: Yes! Agreed!

But in the next PR we will have more checks, so I will keep it as is for now :D
